Allow multiple nonlinear systems for the same FE problem/mesh

lindsayad opened the pull request:
Closes #8780

lindsayad committed:
2500 lines through FEProblemBase

Refs #8780

lindsayad committed:
5000 lines through FEProblemBase

Refs #8780

lindsayad committed:
Completed initial trip through FEProblemBase

Refs #8780

lindsayad committed:
Remove some remnants of PetscTSExecutioner

hugary1995 commented:
I just read through the issue you referenced. It was a fun one to read actually. Though it seems that a few people are against of having multiple systems (mainly due to difficulty in maintenance?). Hopefully the picture has changed in the past few years.

hugary1995 commented:
Also, I'm onboard with this design. I'm happy to provide my use cases when needed.

lindsayad committed:
Make framework compile

Interesting developments
- Generalize code in `UpdateDisplacedMeshThread`

lindsayad committed:
Get moose test to compile

lindsayad committed:
Get moose tests to pass

lindsayad committed:
Compile modules

lindsayad committed:
Respect user LDFLAGS

lindsayad committed:
Make modules tests pass

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1196869/) on 3357ab0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22226/site)

This comment will be updated on new commits.

lindsayad committed:
Fix insidious discrepancy of sorting of pointers in parallel bug

lindsayad committed:
Make apps compile/pass tests

Some important changes:
- Create nlConverged API for backwards compatibility

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1198680/) on 431bc92 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.30%</td>
      <td>84.38%</td>
      <td>+0.08%</td>
      <td>83.91%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80163</td>
      <td>80434</td>
      <td>+271</td>
      <td>1163</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14928</td>
      <td>14885</td>
      <td>-43</td>
      <td>223</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.53%</td>
      <td>89.54%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3285</td>
      <td>3286</td>
      <td>+1</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>384</td>
      <td>384</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/contact)

### External petsc solver

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.48%</td>
      <td>86.48%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>371</td>
      <td>371</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>58</td>
      <td>58</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/external_petsc_solver)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/external_petsc_solver)

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.73%</td>
      <td>89.71%</td>
      <td>-0.02%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3338</td>
      <td>3339</td>
      <td>+1</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>382</td>
      <td>383</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/heat_conduction)

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.37%</td>
      <td>88.37%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>319</td>
      <td>319</td>
      <td>-</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>42</td>
      <td>42</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/level_set)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/level_set)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.01%</td>
      <td>90.01%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12488</td>
      <td>12486</td>
      <td>-2</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1386</td>
      <td>1386</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/phase_field)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.53%</td>
      <td>89.52%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26239</td>
      <td>26237</td>
      <td>-2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3070</td>
      <td>3072</td>
      <td>+2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/tensor_mechanics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>98d9ce</th>
      <th colspan=4>#22226 431bc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.91%</td>
      <td>81.90%</td>
      <td>-0.01%</td>
      <td>88.89%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6346</td>
      <td>6352</td>
      <td>+6</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1402</td>
      <td>1404</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22226/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22226/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 83.91% is less than the suggested 90.0%
- `heat_conduction` new line coverage rate 66.67% is less than the suggested 90.0%
- `xfem` new line coverage rate 88.89% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Revert "Respect user LDFLAGS"

This reverts commit 07b13e535cb430e5b4fe85d645505305e6744f59.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1200332/) on 075eff1 : invalidated by @lindsayad

lindsayad committed:
Test

friedmud commented:
I assume that one of the people that was against it in the past that is for it is me ;-)

When we discussed this last time things were in a very different state.  We had a lot of work to do on MultiApps to get it where we needed it.  It's now settled down quite a bit.  In addition, many new use-cases for this new capability have emerged (including CFD).

As for this PR: It's incredibly well done!  Very well thought out - and well implemented.  Great job @lindsayad !

loganharbour commented:
There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not

lindsayad committed:
Revert "Always use single point map for both residual and Jacobian"

This reverts commit e494367a47fa49302b2bfe06f3f4d33386623833.

lindsayad committed:
Use ADReal::do_derivatives flag when deciding to allow affine opt

And comment about why we only perform the affine map optimizations
when computing a residual

lindsayad committed:
Generalize assignment of displacements to multiple Assembly objects

lindsayad committed:
Generalize derivative insertion checks and test

lindsayad commented:
> There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not

This is an example of where it's appropriate
- we reinit all assembly objects with a given element
- we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any

lindsayad committed:
Go back to always computing the single point map

E.g. forget the affine map optimization. Relying on the affine map
optimization led to multiple failed tests in modules. If anyone ever
shows this as being a hot spot in their simulations, then we can
return and reconsider this optimization

loganharbour commented:
> > There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not
> 
> This is an example of where it's appropriate
> 
>     * we reinit all assembly objects with a given element
> 
>     * we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any

Is there a reason why we wouldn't want to just use the actual system from the primary variable?

loganharbour commented on framework/src/base/Assembly.C:
```diff
@@ -1110,11 +1043,10 @@ Assembly::computeSinglePointMapAD(const Elem * elem,
-        if (computingJacobian())
-          for (const auto & disp_num : _displacements)
+        if (do_derivatives)
+          for (const auto & [disp_num, direction] : _disp_numbers_and_directions)
```
hnnnnng it's beautiful

loganharbour commented:
> > > There's a lot of cases here where we're pulling assembly for system zero; is that appropriate? I feel like this might be hard to track down in the future if it's not
> > 
> > 
> > This is an example of where it's appropriate
> > ```
> > * we reinit all assembly objects with a given element
> > 
> > * we now want to resize something, say stateful material properties. We can pull any assembly's quadrature rule and it should have the same number of quadrature points for resizing. In such an example, pulling at index 0 is as good as any
> > ```
> 
> Is there a reason why we wouldn't want to just use the actual system from the primary variable?
> 
> Edit: oh wait... this is aux

Okay you can ignore this - I'm on board.

Almost through, nothing so far. FEProblemBase is a beast... will finish up this afternoon

lindsayad committed:
Ensure at framework level we don't register coupling across systems

lindsayad committed:
Add share directory to gitignore

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1213671/) on 93e79bb : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1213698/) on 93e79bb : invalidated by @lindsayad

lindsayad commented:
Thank you for the review!!!
